File: takengine/takengine/src/main/java/com/atakmap/map/layer/raster/osm/OSMUtils.java

package com.atakmap.map.layer.raster.osm;

import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.Set;

import com.atakmap.annotations.DeprecatedApi;
import com.atakmap.coremap.log.Log;

import android.database.sqlite.SQLiteDatabase;

import com.atakmap.database.DatabaseIface;
import com.atakmap.database.Databases;
import com.atakmap.database.android.AndroidDatabaseAdapter;

public class OSMUtils {

    public final static String TAG = "OSMUtils";

    public final static String OSMDROID_SQLITE_TILES_TABLE_NAME = "tiles";

    public final static Set<String> OSMDROID_SQLITE_TILES_TABLE_COLUMNS;
    static {
        Set<String> s = new LinkedHashSet<String>();
        s.add("key");
        s.add("provider");
        s.add("tile");
        OSMDROID_SQLITE_TILES_TABLE_COLUMNS = Collections.unmodifiableSet(s);
    }

    private OSMUtils() {
    }

    /** @deprecated use {@link #isOSMDroidSQLite(DatabaseIface)} */
    @Deprecated
    @DeprecatedApi(since = "4.1.1", forRemoval = true, removeAt = "4.4")
    public static boolean isOSMDroidSQLite(SQLiteDatabase database) {
        return isOSMDroidSQLite(new AndroidDatabaseAdapter(database));
    }

    public static boolean isOSMDroidSQLite(DatabaseIface database) {
        try { 
            final Set<String> columns = Databases.getColumnNames(database,
                OSMDROID_SQLITE_TILES_TABLE_NAME);
            if (columns == null)
               return false;
            return columns.equals(OSMDROID_SQLITE_TILES_TABLE_COLUMNS);
        } catch (Exception e) { 
            Log.e(TAG, "bad database encountered: " + database, e);
            return false;
        }
    }

    // next 4 derived from
    // http://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#X_and_Y
    public static double mapnikTileLat(int level, int ytile) {
        final int n = 1 << level;
        return Math.toDegrees(Math.atan(Math.sinh(Math.PI
                * (1.0d - (2.0d * (double) ytile / (double) n)))));
    }

    public static double mapnikTileLng(int level, int xtile) {
        final int n = 1 << level;
        return ((double) xtile / (double) n) * 360.0d - 180.0d;
    }

    public static int mapnikTileX(int level, double lng) {
        return (int)mapnikTileXd(level, lng);
    }

    public static int mapnikTileY(int level, double lat) {
        return (int)mapnikTileYd(level, lat);
    }

    public static double mapnikTileXd(int level, double lng) {
        final int n = 1 << level;
        return (n * ((lng + 180.0d) / 360.0d));
    }

    public static double mapnikTileYd(int level, double lat) {
        final int n = 1 << level;
        final double lat_rad = Math.toRadians(lat);
        return (n*(1d-(Math.log(Math.tan(lat_rad)+(1d/Math.cos(lat_rad)))/Math.PI))/2d);
    }

    public static double mapnikPixelLat(int level, int ytile, int y) {
        return mapnikTileLat(level + 8, (ytile << 8) + y);
    }

    public static double mapnikPixelLng(int level, int xtile, int x) {
        return mapnikTileLng(level + 8, (xtile << 8) + x);
    }

    public static int mapnikPixelY(int level, int ytile, double lat) {
        return mapnikTileY(level + 8, lat) - (ytile << 8);
    }

    public static int mapnikPixelX(int level, int xtile, double lng) {
        return mapnikTileX(level + 8, lng) - (xtile << 8);
    }

    public static double mapnikPixelYd(int level, int ytile, double lat) {
        final double tiley = mapnikTileYd(level, lat);
        return (tiley-ytile)*256d;
    }

    public static double mapnikPixelXd(int level, int xtile, double lng) {
        final double tilex = mapnikTileXd(level, lng);
        return (tilex-xtile)*256d;
    }

    public static double mapnikTileResolution(int level) {
        return mapnikTileResolution(level, 0.0d);
    }

    public static double mapnikTileResolution(int level, double lat) {
        if(level >= 32)
            return 0.0d;
        return 156543.034 * Math.cos(lat) / (1 << level);
    }

    public static int mapnikTileLevel(double resolution) {
        return mapnikTileLevel(resolution, 0.0d);
    }

    public static int mapnikTileLevel(double resolution, double lat) {
        if(resolution == 0.0d)
            return Integer.MAX_VALUE;

        // XXX - not sure whether we want ceil or floor here.....
        return (int)mapnikTileLeveld(resolution, lat);
    }

    public static double mapnikTileLeveld(double resolution, double lat) {
        if(resolution == 0.0d)
            return Integer.MAX_VALUE;
        return Math.log(156543.034 * Math.cos(lat) / resolution) / Math.log(2);
    }

    /**************************************************************************/
    // OSM Droid SQLite

    public static long getOSMDroidSQLiteIndex(long level, long tilex, long tiley) {
        return (((level << level) + tilex) << level) + tiley;
    }

    public static int getOSMDroidSQLiteZoomLevel(long index) {
        for (int i = 0; i < 29; i++)
            if (index <= getOSMDroidSQLiteMaxIndex(i))
                return i;
        return -1;
    }

    public static int getOSMDroidSQLiteTileX(long index) {
        final int zoomLevel = getOSMDroidSQLiteZoomLevel(index);
        if (zoomLevel < 0)
            return -1;
        index >>>= zoomLevel;
        index -= (zoomLevel << zoomLevel);
        return (int) index;
    }

    public static int getOSMDroidSQLiteTileY(long index) {
        final int zoomLevel = getOSMDroidSQLiteZoomLevel(index);
        if (zoomLevel < 0)
            return -1;
        return (int) ((~(0xFFFFFFFFFFFFFFFFL << (long) zoomLevel)) & index);
    }

    public static long getOSMDroidSQLiteMaxIndex(int zoomLevel) {
        return getOSMDroidSQLiteMinIndex(zoomLevel + 1) - 1;

    }

    public static long getOSMDroidSQLiteMinIndex(int zoomLevel) {
        if (zoomLevel < 0 || zoomLevel > 28)
            return -1;
        return (long) zoomLevel << (long) (zoomLevel * 2);
    }
}

File: takkernel/engine/src/main/cpp/raster/osm/OSMUtils.cpp
#include "raster/osm/OSMUtils.h"

#include <climits>
#include <cmath>

#include "math/Utils.h"

namespace atakmap {
    namespace raster{
        namespace osm {

            namespace {
                OSMUtils *singleton = nullptr;
            }

            OSMUtils::OSMUtils() : tableName("tiles")
            {
                colNames.insert("key");
                colNames.insert("provider");
                colNames.insert("tile");
            }

            const OSMUtils& OSMUtils::getInstance()
            {
                if (singleton == nullptr)
                    singleton = new OSMUtils();
                return *singleton;
            }

            const std::string& OSMUtils::getOSMDroidSqliteTilesTableName() const
            {
                return tableName;
            }

            const std::set<std::string>& OSMUtils::getOSMDroidSqliteTilesTableColumns() const
            {
                return colNames;
            }

            bool OSMUtils::isOSMDroidSQLite(db::Database &database) const
            {
                std::vector<TAK::Engine::Port::String> columns =
                    db::getColumnNames(database, tableName.c_str());

                if (columns.size() != colNames.size())
                    return false;

                std::vector<TAK::Engine::Port::String>::iterator iter;
                for (iter = columns.begin(); iter != columns.end(); ++iter) {
                    std::string s(*iter);
                    if (colNames.find(s) == colNames.end())
                        return false;
                }

                return true;
            }

            // next 4 derived from
            // http://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#X_and_Y
            double OSMUtils::mapnikTileLat(int level, int ytile)
            {
                int n = 1 << level;
                return math::toDegrees(atan(sinh(M_PI
                    * (1.0 - (2.0 * (double)ytile / (double)n)))));
            }

            double OSMUtils::mapnikTileLng(int level, int xtile)
            {
                int n = 1 << level;
                return ((double)xtile / (double)n) * 360.0 - 180.0;
            }

            int OSMUtils::mapnikTileX(int level, double lng)
            {
                int n = 1 << level;
                return (int)(n * ((lng + 180.0) / 360.0));
            }

            int OSMUtils::mapnikTileY(int level, double lat)
            {
                int n = 1 << level;
                double lat_rad = math::toRadians(lat);
                return (int)(n
                             * (1.0 - (log(tan(lat_rad) + (1.0 / cos(lat_rad))) / M_PI)) / 2.0);
            }

            double OSMUtils::mapnikPixelLat(int level, int ytile, int y)
            {
                return mapnikTileLat(level + 8, (ytile << 8) + y);
            }

            double OSMUtils::mapnikPixelLng(int level, int xtile, int x)
            {
                return mapnikTileLng(level + 8, (xtile << 8) + x);
            }

            int OSMUtils::mapnikPixelY(int level, int ytile, double lat)
            {
                return mapnikTileY(level + 8, lat) - (ytile << 8);
            }

            int OSMUtils::mapnikPixelX(int level, int xtile, double lng)
            {
                return mapnikTileX(level + 8, lng) - (xtile << 8);
            }

            double OSMUtils::mapnikTileResolution(int level)
            {
                return mapnikTileResolution(level, 0.0);
            }

            double OSMUtils::mapnikTileResolution(int level, double lat)
            {
                if (level >= 32)
                    return 0.0;
                return 156543.034 * cos(lat*M_PI/180.0) / (1 << level);
            }

            int OSMUtils::mapnikTileLevel(double resolution)
            {
                return mapnikTileLevel(resolution, 0.0);
            }

            int OSMUtils::mapnikTileLevel(double resolution, double lat)
            {
                if (resolution == 0.0)
                    return INT_MAX;

                // XXX - not sure whether we want ceil or floor here.....
                return (int)mapnikTileLeveld(resolution, lat);
            }

            double OSMUtils::mapnikTileLeveld(double resolution, double lat)
            {
                if (resolution == 0.0)
                    return INT_MAX;

                return (log(156543.034 * cos(lat) / resolution) / M_LN2);
            }


            /**************************************************************************/
            // OSM Droid SQLite
            int64_t OSMUtils::getOSMDroidSQLiteIndex(int64_t level, int64_t tilex, int64_t tiley)
            {
                return (((level << level) + tilex) << level) + tiley;
            }

            int OSMUtils::getOSMDroidSQLiteZoomLevel(int64_t index)
            {
                for (int i = 0; i < 29; i++)
                    if (index <= getOSMDroidSQLiteMaxIndex(i))
                        return i;
                return -1;
            }

            int OSMUtils::getOSMDroidSQLiteTileX(int64_t index)
            {
                int zoomLevel = getOSMDroidSQLiteZoomLevel(index);
                if (zoomLevel < 0)
                    return -1;
                index >>= zoomLevel;
                index -= (zoomLevel << zoomLevel);
                return (int)index;
            }

            int OSMUtils::getOSMDroidSQLiteTileY(int64_t index)
            {
                int zoomLevel = getOSMDroidSQLiteZoomLevel(index);
                if (zoomLevel < 0)
                    return -1;
                return (int)((~(0xFFFFFFFFFFFFFFFFL << (int64_t)zoomLevel)) & index);
            }

            int64_t OSMUtils::getOSMDroidSQLiteMaxIndex(int zoomLevel)
            {
                return getOSMDroidSQLiteMinIndex(zoomLevel + 1) - 1;
            }

            int64_t OSMUtils::getOSMDroidSQLiteMinIndex(int zoomLevel)
            {
                if (zoomLevel < 0 || zoomLevel > 28)
                    return -1;
                return (int64_t)zoomLevel << (int64_t)(zoomLevel * 2);
            }

        }
    }
}

